// @clr/angular/esm5/utils/focus/key-codes.enum.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
// Event.key key codes for standard browsers and IE/Edge
export var KeyCodes;
(function (KeyCodes) {
    KeyCodes["ArrowUp"] = "ArrowUp";
    KeyCodes["ArrowDown"] = "ArrowDown";
    KeyCodes["ArrowRight"] = "ArrowRight";
    KeyCodes["ArrowLeft"] = "ArrowLeft";
    KeyCodes["Space"] = " ";
    KeyCodes["Escape"] = "Escape";
    KeyCodes["Home"] = "Home";
    KeyCodes["End"] = "End";
    KeyCodes["Enter"] = "Enter";
    KeyCodes["Tab"] = "Tab";
})(KeyCodes || (KeyCodes = {}));
export var IEKeyCodes;
(function (IEKeyCodes) {
    IEKeyCodes["ArrowUp"] = "Up";
    IEKeyCodes["ArrowDown"] = "Down";
    IEKeyCodes["ArrowRight"] = "Right";
    IEKeyCodes["ArrowLeft"] = "Left";
    IEKeyCodes["Space"] = "Spacebar";
    IEKeyCodes["Escape"] = "Esc";
    IEKeyCodes["Home"] = "Home";
    IEKeyCodes["End"] = "End";
    IEKeyCodes["Enter"] = "Enter";
    IEKeyCodes["Tab"] = "Tab";
})(IEKeyCodes || (IEKeyCodes = {}));
