// @clr/angular/esm2015/utils/conditional/if-open.service.js
import * as tslib_1 from "tslib";
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
let IfOpenService = 
/*********
 * @class IfOpenService
 *
 * @description
 * An injectable service used by IfOpen structural directives and the components that implemnt IfOpen in their
 * templates. It holds the value of the open state and provides an Observable that both the directive and the
 * implementing component can subscribe to in order to take action on open value changes.
 *
 */
class IfOpenService {
    constructor() {
        /********
         * @property _openChange
         *
         * @description
         * A RXJS Subject that updates and provides subscriptions to for the current open state of a component template
         * implemting the IfOpen structural directive.
         */
        this._openChange = new Subject();
        /**
         *  Popovers might need to ignore click events on an element
         *  (eg: popover opens on focus on an input field. Clicks should be ignored in this case)
         */
        this._ignoredElementChange = new Subject();
    }
    /*********
     *
     * @description
     * A getter function that provides an observable for the _opened Subject.
     *
     */
    get openChange() {
        return this._openChange.asObservable();
    }
    /*********
     *
     * @description
     * A setter function that updates the current state of _open for this instance of IfOpen structural directive. And,
     * broadcasts the new value to all subscribers.
     *
     * @param value
     */
    set open(value) {
        value = !!value;
        if (this._open !== value) {
            this._open = value;
            this._openChange.next(value);
        }
    }
    /*********
     *
     * @description
     * A getter that returns the current value of this IfOpen instance.
     *
     */
    get open() {
        return this._open;
    }
    toggleWithEvent(event) {
        this.originalEvent = event;
        this.open = !this.open;
        delete this.originalEvent;
    }
    get ignoredElementChange() {
        return this._ignoredElementChange.asObservable();
    }
    registerIgnoredElement(element) {
        this._ignoredElementChange.next(element);
    }
};
IfOpenService = tslib_1.__decorate([
    Injectable()
    /*********
     * @class IfOpenService
     *
     * @description
     * An injectable service used by IfOpen structural directives and the components that implemnt IfOpen in their
     * templates. It holds the value of the open state and provides an Observable that both the directive and the
     * implementing component can subscribe to in order to take action on open value changes.
     *
     */
], IfOpenService);
export { IfOpenService };
