// @clr/angular/esm5/popover/common/popover-old.directive.js
import * as tslib_1 from "tslib";
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Directive, EventEmitter, Input, Output, TemplateRef, ViewContainerRef } from '@angular/core';
import { Point, Popover } from './popover';
var openCount = 0;
var waiting = []; // pending create functions
var PopoverDirectiveOld = /** @class */ (function () {
    function PopoverDirectiveOld(templateRef, viewContainer) {
        this.templateRef = templateRef;
        this.viewContainer = viewContainer;
        this.popoverOptions = {};
        this.clrPopoverOldChange = new EventEmitter(false);
    }
    Object.defineProperty(PopoverDirectiveOld.prototype, "clrPopoverOld", {
        set: function (open) {
            var _this = this;
            if (open) {
                if (this.popoverOptions.allowMultipleOpen) {
                    this.createPopover();
                }
                else {
                    if (openCount === 0) {
                        this.createPopover();
                    }
                    else {
                        waiting.push(function () {
                            _this.createPopover();
                        });
                    }
                }
            }
            else {
                this.viewContainer.clear();
                this.destroyPopover();
                if (!this.popoverOptions.allowMultipleOpen) {
                    if (waiting.length > 0) {
                        var createPopoverFn = waiting.shift();
                        createPopoverFn();
                    }
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    PopoverDirectiveOld.prototype.createPopover = function () {
        var _this = this;
        var embeddedViewRef = this.viewContainer.createEmbeddedView(this.templateRef);
        // TODO: Not sure of the risks associated with using this. Find an alternative.
        // Needed for find the correct height and width of dynamically created views
        // inside of the popover. For Eg: Button Groups
        embeddedViewRef.detectChanges();
        // filter out other nodes in the view ref so we are only left with element nodes
        var elementNodes = embeddedViewRef.rootNodes.filter(function (node) {
            return node.nodeType === 1;
        });
        // we take the first element node in the embedded view; usually there should only be one anyways
        this._popoverInstance = new Popover(elementNodes[0]);
        this._subscription = this._popoverInstance
            .anchor(this.anchorElem, this.anchorPoint, this.popoverPoint, this.popoverOptions)
            .subscribe(function () {
            _this.clrPopoverOldChange.emit(false);
        });
        openCount++;
    };
    PopoverDirectiveOld.prototype.destroyPopover = function () {
        if (this._popoverInstance) {
            this._subscription.unsubscribe();
            this._popoverInstance.release();
            delete this._popoverInstance;
            openCount--;
        }
    };
    PopoverDirectiveOld.prototype.ngOnDestroy = function () {
        this.destroyPopover();
    };
    tslib_1.__decorate([
        Input('clrPopoverOldAnchor'),
        tslib_1.__metadata("design:type", Object)
    ], PopoverDirectiveOld.prototype, "anchorElem", void 0);
    tslib_1.__decorate([
        Input('clrPopoverOldAnchorPoint'),
        tslib_1.__metadata("design:type", Number)
    ], PopoverDirectiveOld.prototype, "anchorPoint", void 0);
    tslib_1.__decorate([
        Input('clrPopoverOldPopoverPoint'),
        tslib_1.__metadata("design:type", Number)
    ], PopoverDirectiveOld.prototype, "popoverPoint", void 0);
    tslib_1.__decorate([
        Input('clrPopoverOldOptions'),
        tslib_1.__metadata("design:type", Object)
    ], PopoverDirectiveOld.prototype, "popoverOptions", void 0);
    tslib_1.__decorate([
        Output('clrPopoverOldChange'),
        tslib_1.__metadata("design:type", Object)
    ], PopoverDirectiveOld.prototype, "clrPopoverOldChange", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], PopoverDirectiveOld.prototype, "clrPopoverOld", null);
    PopoverDirectiveOld = tslib_1.__decorate([
        Directive({ selector: '[clrPopoverOld]' }),
        tslib_1.__metadata("design:paramtypes", [TemplateRef, ViewContainerRef])
    ], PopoverDirectiveOld);
    return PopoverDirectiveOld;
}());
export { PopoverDirectiveOld };
