// @clr/angular/esm2015/layout/tabs/tab-overflow-content.js
import * as tslib_1 from "tslib";
/*
 * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Component, ElementRef, Injector, SkipSelf } from '@angular/core';
import { AbstractPopover } from '../../popover/common/abstract-popover';
import { Point } from '../../popover/common/popover';
let ClrTabOverflowContent = class ClrTabOverflowContent extends AbstractPopover {
    constructor(injector, parentHost) {
        super(injector, parentHost);
        this.anchorPoint = Point.BOTTOM_RIGHT;
        this.popoverPoint = Point.RIGHT_TOP;
        this.closeOnOutsideClick = true;
    }
};
ClrTabOverflowContent = tslib_1.__decorate([
    Component({
        selector: 'clr-tab-overflow-content',
        template: `
        <ng-content></ng-content>
    `,
        host: {
            '[class.dropdown-menu]': 'true',
        }
    }),
    tslib_1.__param(1, SkipSelf()),
    tslib_1.__metadata("design:paramtypes", [Injector, ElementRef])
], ClrTabOverflowContent);
export { ClrTabOverflowContent };
