// @clr/angular/esm5/utils/popover/providers/popover-events.service.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 *
 */
import * as tslib_1 from "tslib";
import { Injectable, Renderer2, Inject } from '@angular/core';
import { ClrPopoverToggleService } from './popover-toggle.service';
import { fromEvent } from 'rxjs';
import { filter } from 'rxjs/operators';
import { DOCUMENT } from '@angular/common';
// https://github.com/angular/angular/issues/20351#issuecomment-344009887
/** @dynamic */
var ClrPopoverEventsService = /** @class */ (function () {
    function ClrPopoverEventsService(renderer, smartOpenService, document) {
        var _this = this;
        this.renderer = renderer;
        this.smartOpenService = smartOpenService;
        this.document = document;
        this.outsideClickClose = true;
        this.scrollToClose = true;
        this.subscriptions = [];
        this.subscriptions.push(smartOpenService.openChange.subscribe(function (open) {
            if (open) {
                _this.addEscapeListener();
                _this.addClickListener();
                _this.addScrollListener();
            }
            else {
                _this.removeAllEventListeners();
            }
        }), smartOpenService.getEventChange().subscribe(function (event) {
            // Remember the event that was used to open the content
            _this.ignoredEvent = event;
        }));
    }
    ClrPopoverEventsService.prototype.addScrollListener = function () {
        var _this = this;
        if (this.scrollToClose) {
            this.documentScroller = fromEvent(this.document, 'scroll', { capture: true });
            this.scrollSubscription = this.documentScroller
                .pipe(filter(this.testForSmartPopoverContentContainer))
                .subscribe(function () {
                _this.smartOpenService.open = false;
                _this.setAnchorFocus();
            });
        }
        else {
            // I think this is where dynamic re-positioning will be added
            // Instead of testing like we do in the close pipe below
            // we need to switch positioning to use an observable and then
            // debounce the scroll events to recalculate content position in a performant way
            // For now, ignore scrolling events.
            return;
        }
    };
    ClrPopoverEventsService.prototype.removeScrollListener = function () {
        if (this.documentScroller) {
            this.scrollSubscription.unsubscribe();
            delete this.documentScroller;
        }
    };
    ClrPopoverEventsService.prototype.testForSmartPopoverContentContainer = function (event) {
        // Filter for the documentScroller observable event targets
        var target = event.target;
        // Walk up the DOM tree until we get to the element that is a direct child of the body.
        while (target.classList && target.parentElement.localName !== 'body') {
            target = target.parentElement;
        }
        // Target is the child element of body where the scroll events originated.
        // Return false and prevent the popover content container from closing for any scroll events inside a popover
        // content container.
        if (target.classList) {
            // check scroll events to see if they are happening in popover content or elsewhere
            return target.classList.contains('clr-popover-content') ? false : true;
        }
        else {
            // prevents it from closing right after first opening
            return false;
        }
    };
    ClrPopoverEventsService.prototype.addClickListener = function () {
        var _this = this;
        if (this.outsideClickClose) {
            this.documentClickListener = this.renderer.listen(this.document, 'click', function (event) {
                if (event === _this.ignoredEvent) {
                    // Here we ignore the opening click event (w/o this content opens and immediately closes.
                    delete _this.ignoredEvent;
                }
                else {
                    _this.smartOpenService.open = false;
                    // Rather than a complex change to the focus trap I put focus on the element that was clicked
                    var clickedElement = event.target;
                    clickedElement.focus();
                }
            });
        }
    };
    ClrPopoverEventsService.prototype.removeClickListener = function () {
        if (this.outsideClickClose) {
            delete this.ignoredEvent;
            if (this.documentClickListener) {
                this.documentClickListener();
                delete this.documentClickListener;
            }
        }
    };
    ClrPopoverEventsService.prototype.addEscapeListener = function () {
        var _this = this;
        this.escapeListener = this.renderer.listen(this.document, 'keydown.escape', function (event) {
            _this.smartOpenService.open = false;
            _this.setAnchorFocus();
        });
    };
    ClrPopoverEventsService.prototype.removeEscapeListener = function () {
        if (this.escapeListener) {
            this.escapeListener();
            delete this.escapeListener;
        }
    };
    Object.defineProperty(ClrPopoverEventsService.prototype, "anchorButtonRef", {
        get: function () {
            return this._anchorButtonRef;
        },
        set: function (ref) {
            this._anchorButtonRef = ref;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ClrPopoverEventsService.prototype, "closeButtonRef", {
        get: function () {
            return this._closeButtonRef;
        },
        set: function (ref) {
            this._closeButtonRef = ref;
        },
        enumerable: true,
        configurable: true
    });
    ClrPopoverEventsService.prototype.setCloseFocus = function () {
        this._closeButtonRef.nativeElement.focus();
    };
    ClrPopoverEventsService.prototype.setAnchorFocus = function () {
        this.anchorButtonRef.nativeElement.focus();
    };
    Object.defineProperty(ClrPopoverEventsService.prototype, "contentRef", {
        get: function () {
            return this._contentRef;
        },
        set: function (host) {
            this._contentRef = host;
        },
        enumerable: true,
        configurable: true
    });
    ClrPopoverEventsService.prototype.removeAllEventListeners = function () {
        this.removeScrollListener();
        this.removeClickListener();
        this.removeEscapeListener();
    };
    ClrPopoverEventsService.prototype.ngOnDestroy = function () {
        this.subscriptions.forEach(function (sub) { return sub.unsubscribe(); });
        this.removeAllEventListeners();
    };
    ClrPopoverEventsService = tslib_1.__decorate([
        Injectable(),
        tslib_1.__param(2, Inject(DOCUMENT)),
        tslib_1.__metadata("design:paramtypes", [Renderer2,
            ClrPopoverToggleService,
            HTMLDocument])
    ], ClrPopoverEventsService);
    return ClrPopoverEventsService;
}());
export { ClrPopoverEventsService };
